### PR TITLE
Support copying buffer to 2d texture array in gl backend, copy with proper texture format

### PR DESCRIPTION
**Pull request by kyren**
Fixes #issue
PR checklist:
- [x] `make` succeeds (on *nix)
- [x] `make reftests` succeeds
- [x] tested examples with the following backends: gl
- [ ] `rustfmt` run on changed code

### CODE REVIEW

File: src/backend/gl/src/queue.rs [hunk 1]
```diff
@@ -598,22 +598,46 @@ impl CommandQueue {
...
+                    glow::TEXTURE_2D => {
+                        gl.bind_texture(glow::TEXTURE_2D, Some(texture));
+                        gl.tex_sub_image_2d_pixel_buffer_offset(
+                            glow::TEXTURE_2D,
+                            r.image_layers.level as _,
+                            r.image_offset.x,
+                            r.image_offset.y,
+                            r.image_extent.width as _,
+                            r.image_extent.height as _,
+                            glow::RGBA,
```
Comment (kvark) [inline src/backend/gl/src/queue.rs]: we should fix the format here... RGBA8 should not be the only format supporting buffer transfers :)

**Comment by kyren** [comment]
> Would you want to fix the formats as well here? I'm fine with landing as is.

Yeah... if it's easy to do it really needs to be fixed, let me see...

**Comment by kyren** [comment]
Well, it wasn't trivial exactly, but it isn't too bad of a change.  Sorry about the PR scope creep again 😅

Also includes a drive-by rename of `TextureType` to `TextureTarget`, as that is actually the correct OpenGL term.

File: src/backend/gl/src/queue.rs [hunk 2]
```diff
@@ -595,32 +595,56 @@ impl CommandQueue {
...
+                            glow::TEXTURE_2D_ARRAY,
+                            r.image_layers.level as _,
+                            r.image_offset.x,
+                            r.image_offset.y,
+                            r.image_layers.layers.start as i32,
+                            r.image_extent.width as _,
+                            r.image_extent.height as _,
+                            r.image_layers.layers.end as i32 - r.image_layers.layers.start as i32,
+                            format,
+                            glow::UNSIGNED_BYTE,
```
Comment (kvark) [inline src/backend/gl/src/queue.rs]: this should also depend on the format, i.e. 16 bit signed formats would use `glow::SHORT`, etc

Comment (kyren) [reply src/backend/gl/src/queue.rs]: OH, you're right it would...

**Comment by kyren** [comment]
I couldn't in good conscience add yet another field to the `CopyBufferToTexture` / `CopyTextureToBuffer` tuple enum variants, so I also changed that...

Arguably, maybe storing the target, format, and pixel data type in the `native::ImageKind::Texture` is too much and you could store something smaller and then compute the appropriate target / format / pixel data type?

File: src/backend/gl/src/command.rs [hunk 3]
```diff
@@ -111,17 +111,31 @@ pub enum Command {
     BindAttribute(n::AttributeDesc, n::RawBuffer, i32, u32),
     //UnbindAttribute(n::AttributeDesc),
     CopyBufferToBuffer(n::RawBuffer, n::RawBuffer, command::BufferCopy),
-    CopyBufferToTexture(n::RawBuffer, n::Texture, n::TextureType, command::BufferImageCopy),
+    CopyBufferToTexture {
```
Comment (kvark) [inline src/backend/gl/src/command.rs]: great change!

File: src/backend/gl/src/command.rs [hunk 4]
```diff
@@ -111,17 +111,31 @@ pub enum Command {
...
+        src_buffer: n::RawBuffer,
+        dst_texture: n::Texture,
+        texture_target: n::TextureTarget,
+        texture_format: n::TextureFormat,
+        pixel_type: n::DataType,
+        command: command::BufferImageCopy,
```
Comment (kvark) [inline src/backend/gl/src/command.rs]: should probably be renamed to "data", since it's not a command

**Comment by kvark** [comment]
@kyren fwiw, storing stuff associated with `Texture` is totally fine, since they aren't meant to be `Copy` anyway.

**Comment by kyren** [comment]
Whoops, sorry I didn't see that you had already approved the PR to be merged!  I interrupted bors 😞

**Pull request merged**